File: leetcode/20.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long ll;

class Solution
{
public:
    bool isValid(string s)
    {
        stack<char> sk;

        for (int i = 0; i < s.length(); i++)
        {
            char ch = s.at(i);
            if (ch == '(')
            {
                sk.push(')');
            }
            else if (ch == '[')
            {
                sk.push(']');
            }
            else if (ch == '{')
            {
                sk.push('}');
            }
            else
            {
                if (sk.size() == 0 || sk.top() != ch)
                {
                    return false;
                }
                else
                {
                    sk.pop();
                }
            }
        }
        return sk.size() == 0;
    }
};

File: python/leetcode/20.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        dct = {
            "(": ")",
            "[": "]",
            "{": "}",
        }

        for si in s:
            if si in dct:
                stack.append(dct[si])
            else:
                if not stack or stack[-1] != si:
                    return False
                stack.pop()
        return len(stack) == 0

if __name__ == "__main__":
    s = "()[]{}"
    res = Solution().isValid(s)
    print(res)